spyder/plugins/editor/widgets/tests/test_editor.py

```diff
@@ -574,5 +576,157 @@ def test_tab_copies_find_to_replace(editor_find_replace_bot, qtbot):
+    editor.set_text('ham\n')
+    editor_stack.autosave.autosave(0)  # call #3, should write
+    assert editor_stack._write_to_file.call_count == 2
+
+
+def test_autosave_does_not_save_new_files(editor_bot, mocker):
+    """Test that autosave() does not save newly created files."""
+    editor_stack, editor = editor_bot
+    editor_stack.data[0].newly_created = True
+    mocker.patch.object(editor_stack, '_write_to_file')
+    editor_stack.autosave.autosave(0)
+    editor_stack._write_to_file.assert_not_called()
+
+
+def test_autosave_does_not_save_after_open(base_editor_bot, mocker):
+    """
+    Test that autosave() does not save files immediately after opening.
+
+    Files should only be autosaved after the user made changes.
+    """
+    editor_stack = base_editor_bot
+    txt = 'spam\n'
+    editor_stack.create_new_editor('ham.py', 'ascii', txt, set_current=True)
+    mocker.patch.object(editor_stack, '_write_to_file')
+    editor_stack.autosave.autosave(0)
+    editor_stack._write_to_file.assert_not_called()
+
+
+def test_autosave_does_not_save_after_reload(base_editor_bot, mocker):
+    """
+    Test that autosave() does not save files immediately after reloading.
+
+    Spyder reload the file if it has changed on disk. In that case, there is
+    no need to autosave because the contents in Spyder are identical to the
+    contents on disk.
+    """
+    editor_stack = base_editor_bot
+    txt = 'spam\n'
+    editor_stack.create_new_editor('ham.py', 'ascii', txt, set_current=True)
+    mocker.patch.object(editor_stack, '_write_to_file')
+    mocker.patch('spyder.plugins.editor.widgets.editor.encoding.read',
+                 return_value=(txt, 'ascii'))
+    print(editor_stack.data[0].editor.document().changed_since_autosave)
+    editor_stack.reload(0)
+    print(editor_stack.data[0].editor.document().changed_since_autosave)
+    editor_stack.autosave.autosave(0)
+    editor_stack._write_to_file.assert_not_called()
+
+
+def test_autosave_updates_name_mapping(editor_bot, mocker, qtbot):
+    """Test that autosave() updates name_mapping."""
+    editor_stack, editor = editor_bot
+    assert editor_stack.autosave.name_mapping == {}
+    mocker.patch.object(editor_stack, '_write_to_file')
+    with qtbot.wait_signal(editor_stack.sig_option_changed) as blocker:
+        editor_stack.autosave.autosave(0)
+    expected = {'foo.py': os.path.join(get_conf_path('autosave'), 'foo.py')}
+    assert editor_stack.autosave.name_mapping == expected
+    assert blocker.args == ['autosave_mapping', expected]
+
+
+def test_autosave_ignores_error(editor_bot, mocker):
+    """Test that autosave() ignores errors when writing to file."""
+    editor_stack, editor = editor_bot
+    mock_write = mocker.patch.object(editor_stack, '_write_to_file')
+    try:
+        mock_write.side_effect = PermissionError
+    except NameError:  # Python 2
+        mock_write.side_effect = IOError
+    editor_stack.autosave.autosave(0)
```

Comment:
Pytest will automatically fail the test if it generates an exception, if that's what you're referring to,

Furthermore, IMO we should not simply hide the exception but give the user a friendly warning dialog that the autosave failed, so they're aware they no longer have safety net is no longer in place and potentially give them warning of an I/O problem with the drive, as their actual main-file saves may also fail especially if they are to the same drives.

Reply:
@jitseniesen?

Reply:
I was not sure about the dialog, because if the failure is persistent then a dialog would be displayed every minute, which is rather annoying. But now that autosave can be turned off, I agree that we should inform the user if the autosave fails.

Reply:
I like that.

Reply:
I agree with @CAM-Gerlach's suggestion too.

Reply:
Done.